grpc/src/main/java/com/linecorp/armeria/server/grpc/HandlerRegistry.java

```diff
@@ -79,33 +81,110 @@ private HandlerRegistry(List<ServerServiceDefinition> services,
         return methods.get(methodName);
     }
 
-    List<ServerServiceDefinition> services() {
+    Map<String, ServerServiceDefinition> services() {
         return services;
     }
 
     Map<String, ServerMethodDefinition<?, ?>> methods() {
         return methods;
     }
 
-    static class Builder {
-        // Store per-service first, to make sure services are added/replaced atomically.
-        private final HashMap<String, ServerServiceDefinition> services =
-                new HashMap<String, ServerServiceDefinition>();
+    static final class Builder {
+        private final List<Entry> entries = new ArrayList<>();
 
         Builder addService(ServerServiceDefinition service) {
-            services.put(service.getServiceDescriptor().getName(), service);
+            entries.add(new Entry(service.getServiceDescriptor().getName(), service, null));
             return this;
         }
 
+        Builder addService(String path, ServerServiceDefinition service) {
+            entries.add(new Entry(normalizePath(path), service, null));
+            return this;
+        }
+
+        Builder addService(String path, ServerServiceDefinition service,
+                            MethodDescriptor<?, ?> methodDescriptor) {
+            entries.add(new Entry(normalizePath(path), service, methodDescriptor));
+            return this;
+        }
+
+        private static String normalizePath(String path) {
+            if (path.isEmpty()) {
+                return path;
+            }
+
+            if (path.charAt(0) == '/') {
+                path = path.substring(1);
+            }
+
+            if (path.isEmpty()) {
+                return path;
+            }
+
+            final int lastCharIndex = path.length() - 1;
+            if (path.charAt(lastCharIndex) == '/') {
+                return path.substring(0, lastCharIndex);
+            } else {
+                return path;
+            }
```

Comment:
Is there any chance a user actually wanted to bind a method at `foo/` instead of `foo` ?

Reply:
I think so

Reply:
Ah, That's a good point. Let me keep the trailing slush.